### PR TITLE
Support write and read of boolean masks in `wcs1d-fits`

### PR DESCRIPTION
**Pull request by dhomeier**
This PR includes a test that was missed before merging #1009 and adds functionality to write `Spectrum1D` masks of `dtype = bool` in WCS1D format, resolving an issue that came up in https://github.com/astropy/specutils/pull/1009#issuecomment-1474424678.
As `bool` or `bit` cannot be written natively to FITS, enabling round-tripping requires encoding the original type, for which there also seems to be no standard. I propose to introduce the header keyword 'BFORM', which is neither defined nor reserved in the FITS standard, and use it analogously to 'TFORMn' in BINTABLE extensions, for which these values are allowed (excerpt):
> `TFORMn` value Description
    ’L’                    Logical
    ’X’                   Bit
    ’B’                   Unsigned byte
    ’I’                    16-bit integer

Here `BFORM` is only used for the 1-bit types, as other formats are already recognised, but it might be added for others for consistency.

### CODE REVIEW

**Comment by dhomeier** [comment]
@rosteen do PRs need extra approval for the CI?

**Comment by rosteen** [comment]
> @rosteen do PRs need extra approval for the CI?

They shouldn't, since you've contributed here before.

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Comment by pllim** [comment]
Actions was having problem. I am going to try to close/re-open.

**Comment by pllim** [comment]
If CI does not kick off still, maybe just wait out the server problems. https://www.githubstatus.com/

**Comment by pllim** [comment]
Not quite there. Looks like flux is written to EXT 0 even when there is mask, which is super confusing. Usually when there are multiple data extensions, EXT 0 is primary. Let's see this example:

```python
import numpy as np
from astropy import units as u
from astropy.io import fits
from astropy.wcs import WCS

from specutils import Spectrum1D

w = WCS(naxis=3)
w.wcs.ctype = 'FREQ', 'DEC--TAN', 'RA---TAN'
w.wcs.set()

sp = Spectrum1D(flux=np.ones((8, 9, 10)) * u.nJy, mask=np.zeros((8, 9, 10), dtype=np.int8), wcs=w)
sp.write("ztmp.fits", format="wcs1d-fits", overwrite=True)
fits.info("ztmp.fits")
```

With this patch, I get this:

```
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  FLUX          1 PrimaryHDU      29   (10, 9, 8)   float64
  1  MASK          1 ImageHDU        11   (10, 9, 8)   uint8 (rescales to int8)
```

But I want this:

```
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     303   ()
  1  FLUX          1 ImageHDU        29   (10, 9, 8)   float64
  2  MASK          1 ImageHDU        11   (10, 9, 8)   uint8 (rescales to int8)
```

**Comment by dhomeier** [comment]
> Not quite there. Looks like flux is written to EXT 0 even when there is mask, which is super confusing. Usually when there are multiple data extensions, EXT 0 is primary. Let's see this example:

Yes, if you want the flux written to a different extension, you need to specify (e.g.) `hdu=1`.
The default has always been to write `flux` to the first available HDU, which for ImageHDU format is the primary HDU.
Mask and uncertainty, if present, are written to the subsequent HDUs per the recent PRs, but flux will be in the same HDU it has always been. I don't quite see how this becomes more confusing when a mask HDU is appended – flux is in HDU 0, whether mask and/or uncertainty are there or not.

**Comment by pllim** [comment]
> flux is in HDU 0, whether mask and/or uncertainty are there or not

But this is not how 90% of the data out there are laid out. I have never seen a real life example where the FLUX is still in primary extension if there are ERR or DQ tied to it.

**Comment by dhomeier** [comment]
> But this is not how 90% of the data out there are laid out. I have never seen a real life example where the FLUX is still in primary extension if there are ERR or DQ tied to it.

The SDSS `spPlate` format e.g., with FLUX in the primary HDU and InvVar in the first extension.
SDSS-I/II `spSpec` also has FLUX in the primary HDU, though with ERR and MASK etc. folded into the same data "cube", and the generic IRAF reader does not even support reading from anything but HDU 0 (but also does not account for any uncertainty or mask data).
APOGEE uses a scheme more similar to what you are describing, but writing flux, StdErr and mask to extensions 1, 2 and 3, but it has its own dedicated loader – however perhaps `wcs1d` should adhere to that order as well, writing uncertainty before mask?
Bottom line, I don't have any robust percentage numbers on how all spectra out there are distributed, but for the basic image format using the primary HDU seems at least as common as the first extension. In any case the default for writing a flux-only spectrum has been `hdu=0` since specutils 1.1, and I don't think it's a good idea to change that now.
OTOH changing the default HDU to 1 depending on the presence of `uncertainty` and/or `mask` seems rather more confusing to me (e.g. what should it be if a mask exists, but is skipped in writing by `mask_name=None`?). But if that is what the majority of users find more logical, so be it...

**Comment by pllim** [comment]
What do the other specutils maintainers think?

> should adhere to that order as well, writing uncertainty before mask?

Yes, I think it is common to have the `SCI`, `ERR`, `DQ` in that order.

**Review by pllim — APPROVED** [review]
I had to pass in extra options (`hdu=1, flux_name="SCI"`) to make an existing test pass over at https://github.com/spacetelescope/jdaviz/pull/2094 but it is not the end of the world.

LGTM. Thanks!

**Review by rosteen — APPROVED** [review]
Code looks fine to me, thanks!

**Comment by pllim** [comment]
Please give me a ping when this ends up in a released version, so I can follow up over at Jdaviz to use this for Cubeviz. Thanks!

**Comment by pllim** [comment]
When is the next release for specutils?
